File: packages/sdk-python/src/agentops/quality/evaluator.py
"""Quality evaluator for AgentOps Python SDK."""

import asyncio
import json
import time
import uuid
from typing import Any

import httpx

from .types import (
    CriterionScore,
    QualityConfig,
    QualityRubric,
    QualityScore,
    DEFAULT_RUBRIC,
)


class QualityEvaluator:
    """LLM-as-judge evaluation engine for quality scoring."""

    def __init__(self, config: QualityConfig | None = None):
        self.config = config or QualityConfig()
        self._pending: dict[str, asyncio.Task[QualityScore]] = {}
        self._queue: list[dict[str, Any]] = []
        self._processing = False

    @property
    def is_enabled(self) -> bool:
        return self.config.enabled

    async def evaluate(
        self,
        event_id: str,
        session_id: str,
        prompt: str,
        response: str,
        context: str | None = None,
    ) -> QualityScore:
        """Evaluate a response quality using LLM-as-judge."""
        if not self.config.enabled:
            raise RuntimeError("Quality evaluation is not enabled")

        # Check sampling rate
        import random
        if random.random() > self.config.sampling_rate:
            return self._create_skipped_score(event_id, session_id, "Skipped due to sampling")

        # Check if already evaluating
        if event_id in self._pending:
            return await self._pending[event_id]

        # Create evaluation task
        task = asyncio.create_task(
            self._do_evaluate(event_id, session_id, prompt, response, context)
        )
        self._pending[event_id] = task

        try:
            return await task
        finally:
            self._pending.pop(event_id, None)

    def queue_evaluation(
        self,
        event_id: str,
        session_id: str,
        prompt: str,
        response: str,
        context: str | None = None,
    ) -> None:
        """Queue an evaluation for async processing."""
        if not self.config.enabled:
            return

        import random
        if random.random() > self.config.sampling_rate:
            return

        self._queue.append({
            "event_id": event_id,
            "session_id": session_id,
            "prompt": prompt,
            "response": response,
            "context": context,
        })
        asyncio.create_task(self._process_queue())

    def get_rubric(self) -> QualityRubric:
        return self.config.rubric

    def set_rubric(self, rubric: QualityRubric) -> None:
        self.config.rubric = rubric

    async def _process_queue(self) -> None:
        if self._processing:
            return
        self._processing = True

        while self._queue:
            batch = self._queue[:self.config.max_concurrent]
            self._queue = self._queue[self.config.max_concurrent:]

            await asyncio.gather(
                *[self.evaluate(**item) for item in batch],
                return_exceptions=True,
            )

        self._processing = False

    async def _do_evaluate(
        self,
        event_id: str,
        session_id: str,
        prompt: str,
        response: str,
        context: str | None,
    ) -> QualityScore:
        start_time = int(time.time() * 1000)

        try:
            evaluation_prompt = self._build_evaluation_prompt(prompt, response, context)
            raw_response = await self._call_judge_model(evaluation_prompt)
            parsed = self._parse_judge_response(raw_response)
            overall_score = self._calculate_overall_score(parsed)

            return QualityScore(
                event_id=str(uuid.uuid4()),
                session_id=session_id,
                overall_score=overall_score,
                criterion_scores=parsed,
                rubric_id=self.config.rubric.id,
                judge_model=self.config.judge_model,
                evaluated_at=int(time.time() * 1000),
                evaluation_duration_ms=int(time.time() * 1000) - start_time,
                raw_response=raw_response,
            )
        except Exception as e:
            return self._create_error_score(
                event_id,
                session_id,
                str(e),
                int(time.time() * 1000) - start_time,
            )

    def _build_evaluation_prompt(
        self,
        prompt: str,
        response: str,
        context: str | None,
    ) -> str:
        criteria_text = "\n".join(
            f"- {c.name} ({c.id}): {c.description}"
            for c in self.config.rubric.criteria
        )

        context_section = f"\n## Context\n{context}\n" if context else ""

        return f"""You are an AI response quality evaluator. Evaluate the following AI response based on the given criteria.

## Criteria
{criteria_text}

## User Prompt
{prompt}
{context_section}
## AI Response
{response}

## Instructions
For each criterion, provide:
1. A score from 1-10 (1=very poor, 10=excellent)
2. A brief reasoning (1-2 sentences)

Respond in JSON format:
{{
  "scores": [
    {{"criterionId": "accuracy", "score": 8, "reasoning": "..."}},
    {{"criterionId": "helpfulness", "score": 7, "reasoning": "..."}},
    ...
  ]
}}"""

    async def _call_judge_model(self, prompt: str) -> str:
        endpoint = self.config.judge_endpoint or "https://api.openai.com/v1/chat/completions"

        async with httpx.AsyncClient(timeout=self.config.timeout_ms / 1000) as client:
            response = await client.post(
                endpoint,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.config.judge_api_key}",
                },
                json={
                    "model": self.config.judge_model,
                    "messages": [
                        {
                            "role": "system",
                            "content": "You are a quality evaluation assistant. Always respond with valid JSON.",
                        },
                        {"role": "user", "content": prompt},
                    ],
                    "temperature": 0.1,
                    "response_format": {"type": "json_object"},
                },
            )
            response.raise_for_status()
            data = response.json()
            return data["choices"][0]["message"]["content"]

    def _parse_judge_response(self, response: str) -> list[CriterionScore]:
        try:
            parsed = json.loads(response)
            scores = parsed.get("scores", [])
            
            return [
                CriterionScore(
                    criterion_id=str(s.get("criterionId", "")),
                    score=max(1, min(10, int(s.get("score", 5)))),
                    reasoning=str(s.get("reasoning", "No reasoning provided")),
                )
                for s in scores
            ]
        except (json.JSONDecodeError, KeyError):
            return [
                CriterionScore(
                    criterion_id=c.id,
                    score=5,
                    reasoning="Unable to parse evaluation response",
                )
                for c in self.config.rubric.criteria
            ]

    def _calculate_overall_score(self, scores: list[CriterionScore]) -> float:
        if not scores:
            return 5.0

        total_weight = 0.0
        weighted_sum = 0.0

        for score in scores:
            criterion = next(
                (c for c in self.config.rubric.criteria if c.id == score.criterion_id),
                None,
            )
            weight = criterion.weight if criterion else 1 / len(scores)
            weighted_sum += score.score * weight
            total_weight += weight

        return round(weighted_sum / total_weight, 1) if total_weight > 0 else 5.0

    def _create_skipped_score(
        self,
        event_id: str,
        session_id: str,
        reason: str,
    ) -> QualityScore:
        return QualityScore(
            event_id=str(uuid.uuid4()),
            session_id=session_id,
            overall_score=0,
            criterion_scores=[],
            rubric_id=self.config.rubric.id,
            judge_model=self.config.judge_model,
            evaluated_at=int(time.time() * 1000),
            evaluation_duration_ms=0,
            error=reason,
        )

    def _create_error_score(
        self,
        event_id: str,
        session_id: str,
        error: str,
        duration_ms: int,
    ) -> QualityScore:
        return QualityScore(
            event_id=str(uuid.uuid4()),
            session_id=session_id,
            overall_score=0,
            criterion_scores=[],
            rubric_id=self.config.rubric.id,
            judge_model=self.config.judge_model,
            evaluated_at=int(time.time() * 1000),
            evaluation_duration_ms=duration_ms,
            error=error,
        )

File: packages/sdk-ts/src/quality/evaluator.ts
/**
 * AgentOps SDK - Quality Evaluator
 *
 * LLM-as-judge evaluation engine for quality scoring.
 */

import type {
  QualityConfig,
  ResolvedQualityConfig,
  QualityScore,
  QualityRubric,
  CriterionScore,
} from "./types.js";
import { DEFAULT_RUBRIC } from "./types.js";
import { generateEventId, now } from "../utils.js";

const DEFAULT_QUALITY_CONFIG: ResolvedQualityConfig = {
  enabled: false,
  judgeModel: "gpt-4o-mini",
  rubric: DEFAULT_RUBRIC,
  samplingRate: 1.0,
  maxConcurrent: 5,
  timeoutMs: 30000,
};

interface EvaluationRequest {
  eventId: string;
  sessionId: string;
  prompt: string;
  response: string;
  context?: string;
}

interface JudgeResponse {
  scores: Array<{
    criterionId: string;
    score: number;
    reasoning: string;
  }>;
}

export class QualityEvaluator {
  private readonly config: ResolvedQualityConfig;
  private pendingEvaluations: Map<string, Promise<QualityScore>> = new Map();
  private evaluationQueue: EvaluationRequest[] = [];
  private isProcessing = false;

  constructor(config?: QualityConfig) {
    this.config = {
      ...DEFAULT_QUALITY_CONFIG,
      ...config,
      rubric: config?.rubric ?? DEFAULT_QUALITY_CONFIG.rubric,
    };
  }

  get isEnabled(): boolean {
    return this.config.enabled;
  }

  /**
   * Evaluate a response quality using LLM-as-judge
   */
  async evaluate(request: EvaluationRequest): Promise<QualityScore> {
    if (!this.config.enabled) {
      throw new Error("Quality evaluation is not enabled");
    }

    // Check sampling rate
    if (Math.random() > this.config.samplingRate) {
      return this.createSkippedScore(request, "Skipped due to sampling");
    }

    // Check if already evaluating
    const existingEvaluation = this.pendingEvaluations.get(request.eventId);
    if (existingEvaluation) {
      return existingEvaluation;
    }

    // Queue the evaluation
    const evaluationPromise = this.doEvaluate(request);
    this.pendingEvaluations.set(request.eventId, evaluationPromise);

    try {
      const result = await evaluationPromise;
      return result;
    } finally {
      this.pendingEvaluations.delete(request.eventId);
    }
  }

  /**
   * Queue an evaluation for async processing
   */
  queueEvaluation(request: EvaluationRequest): void {
    if (!this.config.enabled) return;

    // Check sampling rate
    if (Math.random() > this.config.samplingRate) return;

    this.evaluationQueue.push(request);
    this.processQueue();
  }

  /**
   * Get current rubric
   */
  getRubric(): QualityRubric {
    return this.config.rubric;
  }

  /**
   * Update rubric
   */
  setRubric(rubric: QualityRubric): void {
    (this.config as { rubric: QualityRubric }).rubric = rubric;
  }

  private async processQueue(): Promise<void> {
    if (this.isProcessing) return;
    this.isProcessing = true;

    while (this.evaluationQueue.length > 0) {
      // Process up to maxConcurrent evaluations
      const batch = this.evaluationQueue.splice(0, this.config.maxConcurrent);
      await Promise.allSettled(batch.map((request) => this.evaluate(request)));
    }

    this.isProcessing = false;
  }

  private async doEvaluate(request: EvaluationRequest): Promise<QualityScore> {
    const startTime = now();

    try {
      const prompt = this.buildEvaluationPrompt(request);
      const response = await this.callJudgeModel(prompt);
      const parsed = this.parseJudgeResponse(response);

      const overallScore = this.calculateOverallScore(parsed.scores);

      return {
        eventId: generateEventId(),
        sessionId: request.sessionId,
        overallScore,
        criterionScores: parsed.scores,
        rubricId: this.config.rubric.id,
        judgeModel: this.config.judgeModel,
        evaluatedAt: now(),
        evaluationDurationMs: now() - startTime,
        rawResponse: response,
      };
    } catch (error) {
      return this.createErrorScore(request, error, now() - startTime);
    }
  }

  private buildEvaluationPrompt(request: EvaluationRequest): string {
    const criteriaText = this.config.rubric.criteria
      .map((c) => `- ${c.name} (${c.id}): ${c.description}`)
      .join("\n");

    return `You are an AI response quality evaluator. Evaluate the following AI response based on the given criteria.

## Criteria
${criteriaText}

## User Prompt
${request.prompt}

${request.context ? `## Context\n${request.context}\n` : ""}
## AI Response
${request.response}

## Instructions
For each criterion, provide:
1. A score from 1-10 (1=very poor, 10=excellent)
2. A brief reasoning (1-2 sentences)

Respond in JSON format:
{
  "scores": [
    {"criterionId": "accuracy", "score": 8, "reasoning": "..."},
    {"criterionId": "helpfulness", "score": 7, "reasoning": "..."},
    ...
  ]
}`;
  }

  private async callJudgeModel(prompt: string): Promise<string> {
    const endpoint =
      this.config.judgeEndpoint ?? "https://api.openai.com/v1/chat/completions";

    const controller = new AbortController();
    const timeoutId = setTimeout(
      () => controller.abort(),
      this.config.timeoutMs,
    );

    try {
      const response = await fetch(endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.config.judgeApiKey}`,
        },
        body: JSON.stringify({
          model: this.config.judgeModel,
          messages: [
            {
              role: "system",
              content:
                "You are a quality evaluation assistant. Always respond with valid JSON.",
            },
            { role: "user", content: prompt },
          ],
          temperature: 0.1,
          response_format: { type: "json_object" },
        }),
        signal: controller.signal,
      });

      if (!response.ok) {
        throw new Error(
          `Judge API error: ${response.status} ${response.statusText}`,
        );
      }

      const data = (await response.json()) as {
        choices: Array<{ message: { content: string } }>;
      };
      return data.choices[0]?.message?.content ?? "";
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private parseJudgeResponse(response: string): JudgeResponse {
    try {
      const parsed = JSON.parse(response) as JudgeResponse;

      if (!Array.isArray(parsed.scores)) {
        throw new Error("Invalid response format: missing scores array");
      }

      // Validate and normalize scores
      const validatedScores: CriterionScore[] = parsed.scores.map((s) => ({
        criterionId: String(s.criterionId),
        score: Math.max(1, Math.min(10, Number(s.score) || 5)),
        reasoning: String(s.reasoning || "No reasoning provided"),
      }));

      return { scores: validatedScores };
    } catch (error) {
      // Return default scores on parse error
      return {
        scores: this.config.rubric.criteria.map((c) => ({
          criterionId: c.id,
          score: 5,
          reasoning: "Unable to parse evaluation response",
        })),
      };
    }
  }

  private calculateOverallScore(scores: CriterionScore[]): number {
    let totalWeight = 0;
    let weightedSum = 0;

    for (const score of scores) {
      const criterion = this.config.rubric.criteria.find(
        (c) => c.id === score.criterionId,
      );
      const weight = criterion?.weight ?? 1 / scores.length;
      weightedSum += score.score * weight;
      totalWeight += weight;
    }

    return totalWeight > 0
      ? Math.round((weightedSum / totalWeight) * 10) / 10
      : 5;
  }

  private createSkippedScore(
    request: EvaluationRequest,
    reason: string,
  ): QualityScore {
    return {
      eventId: generateEventId(),
      sessionId: request.sessionId,
      overallScore: 0,
      criterionScores: [],
      rubricId: this.config.rubric.id,
      judgeModel: this.config.judgeModel,
      evaluatedAt: now(),
      evaluationDurationMs: 0,
      error: reason,
    };
  }

  private createErrorScore(
    request: EvaluationRequest,
    error: unknown,
    durationMs: number,
  ): QualityScore {
    const errorMessage = error instanceof Error ? error.message : String(error);

    return {
      eventId: generateEventId(),
      sessionId: request.sessionId,
      overallScore: 0,
      criterionScores: [],
      rubricId: this.config.rubric.id,
      judgeModel: this.config.judgeModel,
      evaluatedAt: now(),
      evaluationDurationMs: durationMs,
      error: errorMessage,
    };
  }
}
